proto/backend.proto

```diff
@@ -0,0 +1,217 @@
+// QUERY
+//-----------------------------------------------
+
+message TimeRange {
+  string fromRaw = 1;
+  string toRaw = 2;
+  int64 fromEpochMs = 3;
+  int64 toEpochMs = 4;
+}
+
+message DataQuery {
+  string refId = 1;
+  int64 maxDataPoints = 2;
+  int64 intervalMs = 3;
+  TimeRange timeRange = 4; // NOTE: this moved from QueryRequest in v1
+  string json = 5; // Arbitrary structure (model?  modelJson in v1)
+}
+
+message QueryRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // List of queries
+  repeated DataQuery queries = 3;
+}
+
+message QueryResponse {
+  // Arrow encoded DataFrames
+  // Each frame encodes its own: Errors, meta, and refId
+  repeated bytes data = 1;
+
+  // Additional response metadata
+  map<string,string> headers = 2;
+}
+
+
+//-----------------------------------------------
+// HTTP style request
+//-----------------------------------------------
+
+message PluginRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // HTTP Style parameters
+  string method = 3; // GET/POST/DELETE/etc
+  string path = 4;   // (with params?)
+  bytes body = 5;    // assume string unless header says otherwise?
+}
+
+message PluginResponse {
+  int32 code = 1;  // Match HTTP status codes
+  bytes body = 2;
+  map<string,string> headers = 3;
+}
+
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Request(PluginRequest) returns (PluginResponse);
+
+  // Well typed query interface
+  rpc Query(QueryRequest) returns (QueryResponse);
+
+  // TODO -- currently queries die when the response >4mb...
+  // We need a better solution...
+  // a) encode paging info in the DataFrame metadata?
+  // b) chunk the results and grafana backend would
+}
+
+
+//-----------------------------------------------
+// Allow plugins to access other plugins
+// Only allow this for some plugins?  GEL only
+//
+// ???? How does the backend know the PluginConfig????
+//-----------------------------------------------
+
+service GrafanaPlatform {
+  rpc PlatformPluginRequest(PluginRequest) returns (PluginResponse);
+
+  rpc PlatformPluginQuery(QueryRequest) returns (QueryResponse);
+}
+
+
+//-----------------------------------------------------------------
+// Streaming support aimed to support similar features to STOMP
+// consider Apache Flight for more complex/standard
+//-----------------------------------------------------------------
+
+message StreamingRequest {
+  int64 pluginId = 1; // match the id from connect request (app or datasource)
+
+  // Environment info
+  map<string,string> headers = 1;
+
+  // The topic we are talking to
```

Comment:
What is a "topic"?

Reply:
This is like a message broker routing topic (used in STOMP, rabbitmq, etc)  essentially it is a path that a user will subscribe to and get all the events that happen to that path,

Digging deeper into spring/rabbitmq, they use the term `destination` and the prefix `topic/*` gets special routing in rabbitmq.  I think reasonable alternative names could be:
* route
* destination
* topic

Reply:
It is essentially a key that every client looking at the same key will get the same messages.  The backend plugin does not manage the broadcast to all subscribers, just push to a topic -- and our websocket library will manage all the subscriptions

Reply:
Implementation wise, each `topic` will get a hub attached:
https://github.com/gorilla/websocket/blob/master/examples/chat/hub.go